In [3]:
from shapely.geometry import Point
import pandas as pd
import pickle
import matplotlib.pyplot as plt
from datetime import date, timedelta, datetime
import geopandas as gpd

loading may data - pickle and csv


In [4]:
%%time
may_loc = pd.read_pickle("../data/may.pkl")
may_trip = pd.read_csv("../data/may_trip.csv")

Wall time: 5.19 s


In [5]:
may_trip['pubTimeStamp'] = pd.to_datetime(may_trip['pubTimeStamp'])


This section makes a dictionary of dataframes. The dataframes can be "chunked up" into different time segments. The main things you need to adjust: the start and end variables, and the delta variable, which designates how "big" a "time chunk" should be.

In [6]:
#make a list of datetimes with datetime function 
start = datetime(2019, 5, 26, 0, 0)
end = datetime(2019, 5, 27, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [7]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, may_trip,"pubTimeStamp")


1


In [8]:
the_output.keys()

dict_keys(['2019-05-26-00-00'])

I'm interested in May 26th specifically, the Saturday of Memorial Day weekend, since it had the most scooter rides.

In [9]:
the_output['2019-05-26-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp    companyName tripRecordNum  \
176114 2019-05-26 00:00:11.587           Lyft          LFT1   
176115 2019-05-26 00:00:24.110           Lyft          LFT2   
176116 2019-05-26 00:00:41.747           Lyft          LFT3   
176117 2019-05-26 00:01:06.000  Bolt Mobility      BOL00985   
176118 2019-05-26 00:01:06.000  Bolt Mobility      BOL00987   
...                        ...            ...           ...   
191193 2019-05-26 23:59:03.500           Lyft       LFT2878   
191194 2019-05-26 23:59:09.507           Lyft       LFT2879   
191195 2019-05-26 23:59:32.867           Lyft       LFT2880   
191196 2019-05-26 23:59:33.527           Lyft       LFT2881   
191197 2019-05-26 23:59:37.390           Lyft       LFT2882   

                                              sumdID  tripDuration  \
176114                                 Powered707225      3.878200   
176115                                 Powered763535     30.441050   
176116                                 Powered759730     18.248550   
176117  Powered-ae44754d-5be7-0c96-4f4f-d106dc6b05ab    567.000000   
176118  Powered-40d728a6-1adb-3e5b-e813-5040890b64b5   1609.000000   
...                                              ...           ...   
191193                                 Powered938052     19.499717   
191194                                 Powered963751      2.989867   
191195                                 Powered658939     48.169433   
191196                                 Powered490647     15.444733   
191197                                 Powered333075      0.206883   

        tripDistance            startDate        startTime  \
176114     314.96064  2019-05-25 00:00:00  23:56:18.766666   
176115    4908.13664  2019-05-25 00:00:00  23:29:57.500000   
176116    8182.41496  2019-05-25 00:00:00  23:42:26.736666   
176117    6102.36000  2019-05-25 00:00:00         23:48:50   
176118   16656.82000  2019-05-25 00:00:00         23:32:38   
...              ...                  ...              ...   
191193    3904.19960  2019-05-26 00:00:00  23:39:33.406666   
191194     790.68244  2019-05-26 00:00:00         23:56:10   
191195   16259.84304  2019-05-26 00:00:00  23:11:22.580000   
191196    5593.83220  2019-05-26 00:00:00  23:44:06.750000   
191197       0.00000  2019-05-26 00:00:00  23:59:24.883333   

                    endDate          endTime  startLatitude  startLongitude  \
176114  2019-05-26 00:00:00  00:00:11.460000      36.148090      -86.810720   
176115  2019-05-26 00:00:00  00:00:23.963333      36.158920      -86.777250   
176116  2019-05-26 00:00:00  00:00:41.650000      36.153780      -86.784050   
176117  2019-05-25 00:00:00         23:58:17      36.149060      -86.806547   
176118  2019-05-25 00:00:00         23:59:27      36.154707      -86.785128   
...                     ...              ...            ...             ...   
191193  2019-05-26 00:00:00  23:59:03.390000      36.164010      -86.771450   
191194  2019-05-26 00:00:00  23:59:09.393333      36.154320      -86.784740   
191195  2019-05-26 00:00:00  23:59:32.746666      36.163660      -86.769230   
191196  2019-05-26 00:00:00  23:59:33.433333      36.149350      -86.798080   
191197  2019-05-26 00:00:00  23:59:37.296666      36.162900      -86.771270   

        endLatitude  endLongitude  \
176114    36.147870    -86.810880   
176115    36.153800    -86.783980   
176116    36.162240    -86.779540   
176117    36.156050    -86.790311   
176118    36.155181    -86.786919   
...             ...           ...   
191193    36.160620    -86.774840   
191194    36.152430    -86.783840   
191195    36.160490    -86.775160   
191196    36.152170    -86.795010   
191197    36.162880    -86.771270   

                                                tripRoute  \
176114  [[36.14809,-86.81072],[36.14805,-86.81075],[36...   
176115  [[36.15892,-86.77725],[36.15901,-86.77723],[36...   
176116  [[36.15378,-86.78405],[36.15452,-86.78459],[36...   
176117  [["36.149060","-86.806547"],[

In [10]:
the_output['2019-05-26-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN          158
PoweredNRVJKLPRUQ7BI     22
PoweredNPIQBPS757EHG     21
PoweredODIJQ7CYI4BHL     21
PoweredFTMUQQPAHOWBB     19
                       ... 
PoweredDSLMP              1
PoweredPL5KM              1
Powered27579              1
Powered1C6W5              1
Powered658939             1
Name: sumdID, Length: 3879, dtype: int64

What I'd like to do is return only the Scooter IDs that had less than 3 rides on this day. Then be able to plot those points to see where they are clustered, because that would tell us where we can reduce the scooter density.

In [11]:
the_output['2019-05-26-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

PoweredW6ZIU            2
PoweredW8BN3            2
Powered6M6BIVIDG7AB7    2
PoweredBHE6Q            2
PoweredUZXB4            2
                       ..
PoweredDSLMP            1
PoweredPL5KM            1
Powered27579            1
Powered1C6W5            1
Powered658939           1
Name: sumdID, Length: 1499, dtype: int64

So now we can see that on 5/26/19, 3878 (subtracting the row for unknowns) unique scooter IDs had rides. Of those 3878 scooters, 1499 only had 1 or 2 rides (not meeting the 3 rides per day threshold). So only 61% of scooters on this day had 3 or more rides.

I'm also curious to know what percentage of scooters are over-performing though. For instance, how many scooters are being used for at least 6 rides per day (twice the stated goal)?

In [12]:
the_output['2019-05-26-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN          158
PoweredNRVJKLPRUQ7BI     22
PoweredNPIQBPS757EHG     21
PoweredODIJQ7CYI4BHL     21
PoweredFTMUQQPAHOWBB     19
                       ... 
Powered603412             6
PoweredPOUWU7US5N3FP      6
Powered2QTOFIQINRR6M      6
PoweredXUL56SIELSIYR      6
PoweredSEOA4HTDBKRQS      6
Name: sumdID, Length: 863, dtype: int64

So 862 unique scooters had twice (or more) the stated rides-per-day goal on this particular day. That's 22% of the scooters.

May 26th was the busiest day of the month, and the Saturday of a holiday weekend... but what about just a regular weekday? Since the aim of this project is supposed to be about getting scooters used by regular Nashville residents, let's see what the data shows us. For May, the median number of rides per day occured on May 16th. So let's run the same process for that day - what percentage of scooters had fewer than the goal of 3 rides per day, and what percentage of scooters had 6 rides or more (twice the stated goal)?

In [13]:
#make a list of datetimes with datetime function 
start = datetime(2019, 5, 16, 0, 0)
end = datetime(2019, 5, 17, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [14]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, may_trip,"pubTimeStamp")


1


In [15]:
the_output.keys()

dict_keys(['2019-05-16-00-00'])

In [16]:
the_output['2019-05-16-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum          sumdID  \
93181 2019-05-16 00:02:13.453        Lyft          LFT1   Powered938154   
93182 2019-05-16 00:02:16.093        Lyft          LFT2   Powered806743   
93183 2019-05-16 00:02:21.057        Lyft          LFT3   Powered437467   
93184 2019-05-16 00:02:39.433        Lyft          LFT5   Powered626167   
93185 2019-05-16 00:02:41.020        Lyft          LFT6   Powered251861   
...                       ...         ...           ...             ...   
97732 2019-05-16 23:59:16.003        Bird       BRD2499    PoweredKJSKW   
97733 2019-05-16 23:59:16.003        Bird       BRD2495  PoweredUNKNOWN   
97734 2019-05-16 23:59:16.003        Bird       BRD2496    Powered2ZSZE   
97735 2019-05-16 23:59:16.003        Bird       BRD2500    Powered4TCGH   
97736 2019-05-16 23:59:16.003        Bird       BRD2504    Powered5I3DR   

       tripDuration  tripDistance            startDate        startTime  \
93181      9.314300    5859.58024  2019-05-15 00:00:00  23:52:54.483333   
93182     12.068167    8553.14988  2019-05-15 00:00:00  23:50:11.886666   
93183     10.370450    2690.28880  2019-05-15 00:00:00  23:51:58.716666   
93184     13.016367    9025.59084  2019-05-15 00:00:00  23:49:38.283333   
93185     12.167017    8500.65644  2019-05-15 00:00:00  23:50:30.886666   
...             ...           ...                  ...              ...   
97732      4.000000    2624.67200  2019-05-16 00:00:00  23:56:18.543333   
97733      0.000000       0.00000  2019-05-16 00:00:00  23:59:09.083333   
97734      4.000000    2959.31768  2019-05-16 00:00:00  23:57:09.720000   
97735      5.000000    1640.42000  2019-05-16 00:00:00  23:55:22.840000   
97736      9.000000    1640.42000  2019-05-16 00:00:00  23:50:24.116666   

                   endDate          endTime  startLatitude  startLongitude  \
93181  2019-05-16 00:00:00  00:02:13.343333       36.16124       -86.78494   
93182  2019-05-16 00:00:00  00:02:15.976666       36.16033       -86.77564   
93183  2019-05-16 00:00:00  00:02:20.943333       36.15971       -86.77992   
93184  2019-05-16 00:00:00  00:02:39.263333       36.16034       -86.77685   
93185  2019-05-16 00:00:00  00:02:40.906666       36.16017       -86.77623   
...                    ...              ...            ...             ...   
97732  2019-05-17 00:00:00  00:00:09.516666       36.16040       -86.77510   
97733  2019-05-16 00:00:00  23:59:19.866666       36.16140       -86.77750   
97734  2019-05-17 00:00:00  00:01:04.803333       36.17830       -86.79020   
97735  2019-05-17 00:00:00  00:00:19.566666       36.16260       -86.77770   
97736  2019-05-16 00:00:00  23:59:38.173333       36.17630       -86.75610   

       endLatitude  endLongitude  \
93181     36.15180     -86.78417   
93182     36.14548     -86.79282   
93183     36.16007     -86.77317   
93184     36.14546     -86.79266   
93185     36.14548     -86.79265   
...            ...           ...   
97732     36.15480     -86.77480   
97733     36.16140     -86.77750   
97734     36.17150     -86.78570   
97735     36.16080     -86.77880   
97736     36.17820     -86.75150   

                                               tripRoute  \
93181  [[36.16124,-86.78494],[36.16133,-86.78504],[36...   
93182  [[36.16033,-86.77564],[36.16028,-86.77574],[36...   
93183  [[36.15971,-86.77992],[36.15971,-86.77986],[36...   
93184  [[36.16034,-86.77685],[36.1602,-86.77672],[36....   
93185  [[36.16017,-86.77623],[36.16013,-86.77618],[36...   
...                                                  ...   
97732    [[36.160351,-86.775028],[36.154864,-86.774738]]   
97733                                                 []   
97734  [[36.178265,-86.790047],[36.178265,-86.790039]...   
97735  [[36.162796,-86.777749],[36.160911,-86.777488]...   
97736  [[36.176512,-86.756035],[36.176626,-86.755734]...   

                        create_dt  
93181  2019-05-17 17:13:03.513000  
93182  2019-05-17 17:13:03.547000  
93183  2019-05-1

In [17]:
the_output['2019-05-16-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredIOZGHYOGSO7HK    28
PoweredSDQ6GAHIJK5AF    28
Powered5VHFZMDPGBWD5    27
PoweredEE2HRZ3ZRFM5O    27
PoweredJ7LPOP2NKDWYU    27
                        ..
PoweredTSXEH             1
Powered224310            1
Powered794954            1
Powered72PTV             1
Powered5I3DR             1
Name: sumdID, Length: 1665, dtype: int64

In [18]:
the_output['2019-05-16-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

PoweredIFR8I      2
PoweredJXM5N      2
Powered668125     2
PoweredSXRSH      2
Powered9050644    2
                 ..
PoweredTSXEH      1
Powered224310     1
Powered794954     1
Powered72PTV      1
Powered5I3DR      1
Name: sumdID, Length: 1153, dtype: int64

So out of 1665 unique scooters that ran on 5/16/19, 1153 of them had fewer than 3 rides. So only 31% of the scooters met the threshold of 3 rides that day.

In [19]:
the_output['2019-05-16-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredIOZGHYOGSO7HK    28
PoweredSDQ6GAHIJK5AF    28
Powered5VHFZMDPGBWD5    27
PoweredEE2HRZ3ZRFM5O    27
PoweredJ7LPOP2NKDWYU    27
                        ..
PoweredCKT7IMUCI7EVD     6
Powered861765            6
Powered924486            6
Powered590962            6
Powered889438            6
Name: sumdID, Length: 137, dtype: int64

137 scooters, or 8%, performed well with at least 6 rides that day.

In [20]:
#make a list of datetimes with datetime function 
start = datetime(2019, 5, 9, 0, 0)
end = datetime(2019, 5, 10, 0, 0)
delta = timedelta(days=1)
timelist = []
timenamelist = []
while start <= end:
    timelist.append(start)
    timenamelist.append(start.strftime("%Y-%m-%d-%H-%M"))
    start += delta
# print(timelist)
# print(timenamelist)

In [21]:
#goal of function: using list of time frames, in this case, days, make a dataframe containing
#info about the location for each scooter, for each day


#make dictionary of dataframes using base dataframe and a user-defined segments_of_time_list
#from cell above
def make_day_df(datetime_list, timenames_list, dataframe, colname):
    resultslist = {} #turn into dictionary
    x = 0
    y = 1
    z = 0
    while x in range(0,len(datetime_list)-1):
        starttime = datetime_list[x]
        endtime = datetime_list[y]
        mask = (dataframe[colname] > starttime) & (dataframe[colname] <= endtime)
        framename = dataframe.loc[mask]
        x += 1
        y += 1
        print(x)
        resultslist[timenames_list[z]] = framename
        z += 1
    return resultslist
        
the_output = make_day_df(timelist, timenamelist, may_trip,"pubTimeStamp")


1


In [22]:
the_output.keys()

dict_keys(['2019-05-09-00-00'])

In [23]:
the_output['2019-05-09-00-00']
# the_output['2019-05-01-00-00']['sumdid'].nunique()

pubTimeStamp companyName tripRecordNum                sumdID  \
52368 2019-05-09 00:00:13.297        Lyft          LFT1         Powered815197   
52369 2019-05-09 00:00:18.797        Bird       BRD2140          PoweredPW8Q8   
52370 2019-05-09 00:00:18.797        Bird       BRD2139          PoweredQ2UYP   
52371 2019-05-09 00:00:18.797        Bird       BRD2141          PoweredIQYHJ   
52372 2019-05-09 00:00:18.797        Bird       BRD2133          PoweredG7XBR   
...                       ...         ...           ...                   ...   
56675 2019-05-09 23:59:13.410        Lime       LIM2326  PoweredH4GQZH2O546WG   
56676 2019-05-09 23:59:13.410        Lime       LIM2327  Powered6EA2IC57NGEBA   
56677 2019-05-09 23:59:13.410        Lime       LIM2324  PoweredAKWXXWK3HB2WC   
56678 2019-05-09 23:59:13.410        Lime       LIM2325  PoweredKHQWLTETRKMER   
56679 2019-05-09 23:59:13.413        Lime       LIM2328  Powered2PZHBAB7ENUL4   

       tripDuration  tripDistance            startDate        startTime  \
52368      2.602683    2480.31504  2019-05-08 00:00:00         23:57:37   
52369     26.000000    9186.35200  2019-05-08 00:00:00  23:35:58.393333   
52370     25.000000    9842.52000  2019-05-08 00:00:00  23:37:14.716666   
52371     36.000000   14763.78000  2019-05-08 00:00:00  23:25:24.960000   
52372      2.000000       0.00000  2019-05-08 00:00:00  23:58:21.036666   
...             ...           ...                  ...              ...   
56675      8.766667     489.81360  2019-05-09 00:00:00         23:48:15   
56676     31.983333    1062.83760  2019-05-09 00:00:00         23:25:08   
56677      2.133333      97.23120  2019-05-09 00:00:00         23:53:23   
56678     15.833333    1065.27600  2019-05-09 00:00:00         23:40:39   
56679      5.866667     457.80960  2019-05-09 00:00:00         23:53:12   

                   endDate          endTime  startLatitude  startLongitude  \
52368  2019-05-09 00:00:00  00:00:13.160000      36.176140      -86.785500   
52369  2019-05-09 00:00:00  00:02:02.410000      36.149600      -86.811700   
52370  2019-05-09 00:00:00  00:01:57.903333      36.170800      -86.783400   
52371  2019-05-09 00:00:00  00:01:19.546666      36.177500      -86.750200   
52372  2019-05-09 00:00:00  00:00:42.730000      36.136900      -86.791800   
...                    ...              ...            ...             ...   
56675  2019-05-09 00:00:00         23:57:01      36.150968      -86.796438   
56676  2019-05-09 00:00:00         23:57:07      36.171958      -86.764933   
56677  2019-05-09 00:00:00         23:55:31      36.164395      -86.780417   
56678  2019-05-09 00:00:00         23:56:29      36.122852      -86.776075   
56679  2019-05-09 00:00:00         23:59:04      36.150844      -86.800527   

       endLatitude  endLongitude  \
52368    36.170580    -86.782480   
52369    36.148400    -86.815700   
52370    36.170100    -86.784800   
52371    36.198700    -86.722700   
52372    36.137400    -86.791800   
...            ...           ...   
56675    36.160035    -86.779243   
56676    36.160659    -86.775197   
56677    36.164887    -86.779011   
56678    36.140741    -86.766382   
56679    36.144162    -86.813830   

                                               tripRoute  \
52368  [[36.17614,-86.7855],[36.17611,-86.78545],[36....   
52369  [[36.149787,-86.811626],[36.149728,-86.811601]...   
52370  [[36.170729,-86.783374],[36.170649,-86.783351]...   
52371  [[36.177478,-86.750178],[36.177492,-86.750055]...   
52372    [[36.137266,-86.791739],[36.137462,-86.791746]]   
...                                                  ...   
56675  [[36.1527,-86.79454],[36.15411,-86.79290999999...   
56676  [[36.17104,-86.76421],[36.16997,-86.76349],[36...   
56677  [[36.16454,-86.78073],[36.16424,-86.78116],[36...   
56678  [[36.12302,-86.77598],[36.122789999999995,-86....   
56679  [[36.15069,-86.80045],[36.150439999999996,-86....   

                        create_dt  
52368  2019-05-1

In [24]:
the_output['2019-05-09-00-00']['sumdID'].value_counts()
# the_output['2019-05-01-00-00']['sumdid'].nunique()

PoweredUNKNOWN          13
PoweredKQ4EXMRGULEHF    11
Powered5WU2V3OC3WOIG    11
PoweredJUPYKIC36OSG6    11
PoweredDD3XJEU6PEMBV    11
                        ..
PoweredXIBE4             1
PoweredLEQPG             1
PoweredBEXMR             1
PoweredJYRT7             1
PoweredFWNF8             1
Name: sumdID, Length: 1750, dtype: int64

1737 rides total

In [26]:
the_output['2019-05-09-00-00']['sumdID'].value_counts().loc[lambda x : x<3]

PoweredOCJNGRPLCOVVB    2
PoweredJLTNICE7DQVPU    2
Powered7V1G3            2
Powered251145           2
PoweredY26FCKAMWNV34    2
                       ..
PoweredXIBE4            1
PoweredLEQPG            1
PoweredBEXMR            1
PoweredJYRT7            1
PoweredFWNF8            1
Name: sumdID, Length: 1096, dtype: int64

654 scooters had 3 or more rides, or 38%

In [27]:
the_output['2019-05-09-00-00']['sumdID'].value_counts().loc[lambda x : x>5]

PoweredUNKNOWN          13
PoweredKQ4EXMRGULEHF    11
Powered5WU2V3OC3WOIG    11
PoweredJUPYKIC36OSG6    11
PoweredDD3XJEU6PEMBV    11
                        ..
PoweredVUJAJ6H4VVDJF     6
Powered957810            6
Powered889824            6
PoweredSINZNV4OYDROU     6
Powered645069            6
Name: sumdID, Length: 129, dtype: int64

129 scooters had 6 or more rides, or 7%